lxd/seccomp/seccomp.go

```diff
@@ -1856,6 +1885,188 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
+			return 0
+		}
+
+		return int(-C.EFAULT)
+	}
+
+	paramValues := C.GoString(&cBuf[0])
+
+	// disallow specifying module parameters, it's just not safe
+	if string(paramValues) != "" {
+		ctx["err"] = "Forbid setting module parameters"
+		if s.s.OS.SeccompListenerContinue {
+			ctx["syscall_continue"] = "true"
+			C.seccomp_notify_update_response(siov.resp, 0, C.uint32_t(seccompUserNotifFlagContinue))
+			return 0
+		}
+
+		return int(-C.EPERM)
+	}
+
+	// int flags
+	flags := C.int(siov.req.data.args[2])
+
+	// disallow specifying flags, it's just not safe
+	if flags != 0 {
+		ctx["err"] = "Forbid setting flags in finit_module()"
+		if s.s.OS.SeccompListenerContinue {
+			ctx["syscall_continue"] = "true"
+			C.seccomp_notify_update_response(siov.resp, 0, C.uint32_t(seccompUserNotifFlagContinue))
+			return 0
+		}
+
+		return int(-C.EPERM)
+	}
+
+	containerInitPID := int(siov.msg.init_pid)
+	targetPID := int(siov.req.pid)
+
+	ctInitUserNS, err := os.Readlink(fmt.Sprintf("/proc/%d/ns/user", containerInitPID))
+	if err != nil {
+		ctx["err"] = "Can't get userns for ct init process"
+		return int(-C.EPERM)
+	}
+
+	reqUserNS, err := os.Readlink(fmt.Sprintf("/proc/%d/ns/user", targetPID))
+	if err != nil {
+		ctx["err"] = "Can't get userns for requestor process"
+		return int(-C.EPERM)
+	}
+
+	ctx["param_values"] = string(paramValues)
+
+	// nested container?
+	if ctInitUserNS != reqUserNS {
+		ctx["err"] = "finit_module() from nested container is forbidden"
+		return int(-C.EPERM)
+	}
+
+	// has CAP_SYS_MODULE capability?
+	if !C.lxd_pid_cap_is_set(C.int(targetPID), C.CAP_SYS_MODULE, C.CAP_EFFECTIVE) {
+		ctx["err"] = "requestor process creds lacks CAP_SYS_MODULE"
+		return int(-C.EPERM)
+	}
+
+	moduleFileFD, err := unix.Openat(siov.procFd, fmt.Sprintf("fd/%d", fd), unix.O_RDONLY|unix.O_CLOEXEC, 0)
+	if err != nil {
+		ctx["err"] = fmt.Sprintf("Can't open module file (unix.Openat): %s", err)
+		return int(-C.EPERM)
+	}
+
+	moduleFile := os.NewFile(uintptr(moduleFileFD), "/proc/<pid>/fd/<fd>")
+	if moduleFile == nil {
+		ctx["err"] = fmt.Sprintf("Can't open module file (os.NewFile): %s", err)
+		return int(-C.EPERM)
+	}
+
+	defer func() { _ = moduleFile.Close() }()
+
+	forksyscallgoargs := []string{
+		"forksyscallgo",
+		"finit_module_parse",         // <syscall_operation>
+		fmt.Sprintf("%d", targetPID), // <PID>
+		fmt.Sprintf("%d", 0),         // <PidFd>
+		fmt.Sprintf("%d", 3),         // <module_fd>
+	}
+
+	var buffer bytes.Buffer
+	var output bytes.Buffer
+	p := subprocess.NewProcessWithFds(util.GetExecPath(), forksyscallgoargs, nil,
+		&nullWriteCloser{&output}, &nullWriteCloser{&buffer})
+
+	// We want to drop privileges, as we use "debug/elf" Golang
+	// package which is considered as not safe for production use.
+	// By spawning an unprivileged process we can safely use this package.
+	p.SetCreds(s.s.OS.UnprivUID, s.s.OS.UnprivGID)
+
+	// let's set timeout too
```

Comment:
Would be good to explain *why* this is here too.

Reply:
Done!